libyul/optimiser/UnusedFunctionsCommon.h

```diff
@@ -0,0 +1,171 @@
+{
+	// We skip the function body if it
+	// 1. is empty, or
+	// 2. is a single statement that is an assignment statement whose value is a non-builtin
+	//    function call, or
+	// 3. is a single expression-statement that is a non-builtin function call.
+	// The above cases are simple enough so that the inliner alone can remove the parameters.
+	if (_f.body.statements.empty())
+		return true;
+	if (_f.body.statements.size() == 1)
+	{
+		Statement const& e = _f.body.statements[0];
+		if (std::holds_alternative<Assignment>(e))
+		{
+			if (std::holds_alternative<FunctionCall>(*std::get<Assignment>(e).value))
+			{
+				FunctionCall c = std::get<FunctionCall>(*std::get<Assignment>(e).value);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+		}
+		else if (std::holds_alternative<ExpressionStatement>(e))
+			if (std::holds_alternative<FunctionCall>(std::get<ExpressionStatement>(e).expression))
+			{
+				FunctionCall c = std::get<FunctionCall>(std::get<ExpressionStatement>(e).expression);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+	}
+	return false;
+}
+
+FunctionDefinition createReplacement(
+	FunctionDefinition& _old,
+	std::map<YulString, std::vector<bool>> const& _unusedParameters,
+	std::map<YulString, std::vector<bool>> const& _unusedReturnVariables,
+	NameDispenser& _nameDispenser,
+	std::map<YulString, YulString> _inverseTranslations
+)
+{
+	auto generateName = [&](TypedName t)
+	{
+		return TypedName{
+			t.location,
+			_nameDispenser.newName(t.name),
+			t.type
+		};
+	};
+
+	langutil::SourceLocation loc = _old.location;
+	YulString newName = _inverseTranslations.at(_old.name);
+	TypedNameList functionArguments;
+	TypedNameList returnVariables;
+	TypedNameList renamedParameters = util::applyMap(_old.parameters, generateName);
+	TypedNameList reducedRenamedParameters;
+	TypedNameList renamedReturnVariables = util::applyMap(_old.returnVariables, generateName);
+	TypedNameList reducedRenamedReturnVariables;
+
+	if (_unusedParameters.count(newName))
+	{
+		std::vector<bool> const& mask = _unusedParameters.at(newName);
+		functionArguments = applyBooleanMask(_old.parameters, mask);
+		reducedRenamedParameters = applyBooleanMask(renamedParameters, mask);
+	}
+	else
+	{
+		functionArguments = _old.parameters;
+		reducedRenamedParameters = renamedParameters;
+	}
+
+	if (_unusedReturnVariables.count(newName))
+	{
+		std::vector<bool> const& mask = _unusedReturnVariables.at(newName);
+		returnVariables = applyBooleanMask(_old.returnVariables, mask);
+		reducedRenamedReturnVariables = applyBooleanMask(renamedReturnVariables, mask);
+	}
+	else
+	{
+		returnVariables = _old.returnVariables;
+		reducedRenamedReturnVariables = renamedReturnVariables;
+	}
+
+	FunctionDefinition newFunction{
+		loc,
+		std::move(newName),
+		std::move(functionArguments),
+		std::move(returnVariables),
+		{loc, {}} // body
+	};
+
+	std::swap(newFunction.body, _old.body);
```

Comment:
I'm wondering if it would be simpler to modify the parameter list of the original function instead

Reply:
What do you mean?

Reply:
Instead of creating a new function and swapping the body, keep the old function, change its name and prune its parameters. Then add a new function that calls the old.